src/main/scala/com/fulcrumgenomics/bam/AddReadGroupsByName.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.bam
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{Io, ProgressLogger}
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import htsjdk.samtools.SAMFileHeader.SortOrder
+import htsjdk.samtools._
+import htsjdk.samtools.util.Iso8601Date
+
+import scala.collection.mutable
+
+@clp(group = ClpGroups.SamOrBam, description =
+  """
+    |Adds read groups to a BAM file for a single sample by parsing the read name.
+    |
+    |Will add one or read groups by parsing the read names.  Assumes the read names are of the form:
+    |  <instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<xpos>:<y-pos> <read>:<is filtered>:<control number>:<barcode sequence>
+    |
+    |Each unique combination of <instrument>:<run number>:<flowcell ID>:<lane> will be its own read group. The ID of the
+    |read group will be an integer, the platform unit will be <flowcell-id>.<lane>, and the platform model if not given
+    |will be <instrument.run-number>.
+    |
+    |The input is assumed to contain reads for one sample and library.  Therefore, the sample and library must be given
+    |and will be applied to all read groups.
+    |
+    |Two passes will be performed on the input: first to gather all the read groups, and second to write the output BAM
+    |file.
+  """
```

Comment:
I think some mention of what happens if there are read groups in the BAM file already would be helpful (i.e. they are all discarded and replaced).